File: scripts/gen_yasmin_csv.py
#!/usr/bin/env python3
"""Generate CSV of all Yasmin (user_id=139) submissions from the database."""
import csv
import os
import psycopg2

DATABASE_URL = None
# Try to read from next-app/.env.production
env_file = "/home/ubuntu/icpchue/next-app/.env.production"
if os.path.exists(env_file):
    with open(env_file) as f:
        for line in f:
            if line.startswith("DATABASE_URL="):
                DATABASE_URL = line.strip().split("=", 1)[1].strip('"').strip("'")
                break

if not DATABASE_URL:
    print("ERROR: Could not find DATABASE_URL")
    exit(1)

conn = psycopg2.connect(DATABASE_URL)
cur = conn.cursor()

cur.execute("""
    SELECT 
        id, user_id, source, contest_id, problem_index, verdict,
        time_ms, memory_kb, language, paste_events, tab_switches,
        time_to_solve_seconds, attempt_number, cf_submission_id, cf_handle,
        submitted_at, source_code
    FROM submissions
    WHERE user_id = 139 AND source = 'codeforces'
    ORDER BY submitted_at ASC
""")

rows = cur.fetchall()
columns = [desc[0] for desc in cur.description]

output_path = "/home/ubuntu/icpchue/yasmin_submissions.csv"
with open(output_path, "w", newline="", encoding="utf-8") as f:
    writer = csv.writer(f)
    writer.writerow(columns)
    for row in rows:
        writer.writerow(row)

print(f"Written {len(rows)} submissions to {output_path}")
cur.close()
conn.close()

File: scripts/gen_yasmin_csv.js
const { Pool } = require('pg');
const fs = require('fs');
const path = require('path');

// Read DATABASE_URL from .env.production
const envContent = fs.readFileSync('/home/ubuntu/icpchue/next-app/.env.production', 'utf8');
const match = envContent.match(/^DATABASE_URL=(.+)$/m);
if (!match) { console.error('No DATABASE_URL found'); process.exit(1); }
let dbUrl = match[1].replace(/^["']|["']$/g, '');
if (dbUrl.includes('sslmode=require')) {
    dbUrl = dbUrl.replace('?sslmode=require', '').replace('&sslmode=require', '');
}

process.env.NODE_TLS_REJECT_UNAUTHORIZED = '0';
const pool = new Pool({ connectionString: dbUrl, ssl: true });

async function main() {
    const result = await pool.query(`
        SELECT id, user_id, source, contest_id, problem_index, verdict,
               time_ms, memory_kb, language, paste_events, tab_switches,
               time_to_solve_seconds, attempt_number, cf_submission_id, cf_handle,
               submitted_at, source_code
        FROM submissions
        WHERE user_id = 139 AND source = 'codeforces'
        ORDER BY submitted_at ASC
    `);

    const columns = ['id','user_id','source','contest_id','problem_index','verdict','time_ms','memory_kb','language','paste_events','tab_switches','time_to_solve_seconds','attempt_number','cf_submission_id','cf_handle','submitted_at','source_code'];
    
    function escapeCsv(val) {
        if (val === null || val === undefined) return '';
        const str = String(val);
        if (str.includes(',') || str.includes('"') || str.includes('\n') || str.includes('\r')) {
            return '"' + str.replace(/"/g, '""') + '"';
        }
        return str;
    }

    let csv = columns.join(',') + '\n';
    for (const row of result.rows) {
        const line = columns.map(col => escapeCsv(row[col])).join(',');
        csv += line + '\n';
    }

    const outPath = '/home/ubuntu/icpchue/yasmin_submissions.csv';
    fs.writeFileSync(outPath, csv, 'utf8');
    console.log(`Written ${result.rows.length} rows to ${outPath}`);
    await pool.end();
}

main().catch(e => { console.error(e); process.exit(1); });
